src/arkode/arkode_lsrkstep_io.c

```diff
@@ -0,0 +1,661 @@
+  Exported optional output functions.
+  ===============================================================*/
+
+/*---------------------------------------------------------------
+  LSRKStepGetNumRhsEvals:
+
+  Returns the current number of calls to f
+  ---------------------------------------------------------------*/
+int LSRKStepGetNumRhsEvals(void* arkode_mem, int num_rhs_fn, long int* f_evals)
+{
+  ARKodeMem ark_mem;
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMem and ARKodeLSRKStepMem structures */
+  retval = lsrkStep_AccessARKODEStepMem(arkode_mem, __func__, &ark_mem,
+                                        &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get values from step_mem */
+  *f_evals = step_mem->nfe;
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  LSRKStepGetNumDomEigUpdates:
+
+  Returns the number of dominant eigenvalue updates
+  ---------------------------------------------------------------*/
+int LSRKStepGetNumDomEigUpdates(void* arkode_mem, long int* num_dom_eig_updates)
+{
+  ARKodeMem ark_mem;
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMem and ARKodeLSRKStepMem structures */
+  retval = lsrkStep_AccessARKODEStepMem(arkode_mem, __func__, &ark_mem,
+                                        &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get values from step_mem */
+  *num_dom_eig_updates = step_mem->num_dom_eig_updates;
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  LSRKStepGetMaxNumStages:
+
+  Returns the max number of stages used
+  ---------------------------------------------------------------*/
+int LSRKStepGetMaxNumStages(void* arkode_mem, int* stage_max)
+{
+  ARKodeMem ark_mem;
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMem and ARKodeLSRKStepMem structures */
+  retval = lsrkStep_AccessARKODEStepMem(arkode_mem, __func__, &ark_mem,
+                                        &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get values from step_mem */
+  *stage_max = step_mem->stage_max;
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  LSRKStepGetAverageStageNum:
+
+  Returns the average number of stages taken
+  ---------------------------------------------------------------*/
+int LSRKStepGetAverageStageNum(void* arkode_mem, sunrealtype* avg_stage)
+{
+  ARKodeMem ark_mem;
+  ARKodeLSRKStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMem and ARKodeLSRKStepMem structures */
+  retval = lsrkStep_AccessARKODEStepMem(arkode_mem, __func__, &ark_mem,
+                                        &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* get values from step_mem */
+  *avg_stage = ((sunrealtype)step_mem->nfe) /
+               ((sunrealtype)ark_mem->nst_attempts);
```

Comment:
Is this always true? I suppose for low order method the Hermite interpolation would not add extra `fe` evaluations from dense output and root finding shouldn't add any. If there are recoverable RHS failures in a step attempt that would throw this off.

Reply:
I think that is an incredibly low probability corner case, and that this "get" function doesn't need to worry about it.  If anything, @maggul could add a comment to the documentation regarding how the value is computed, and that it would be slightly off if the user RHS function returns recoverable failures.

Reply:
I see this as purely a convenience feature, since users can access all of the relevant information via other means.  I definitely think that it should be logged at each time step, since it could be very interesting to see how this changes for strongly nonlinear applications.

Given the fact that the statistic is so non-critical, I recommend that this function either be retained as-is (with documentation clarifying that it is a convenience routine and explaining how the value is calculated), or that it be removed entirely.